lib/csv_helper/loader.rb

```diff
@@ -57,18 +72,30 @@ def merge_options(options)
 
     # Default validations are run during import, which prevent bad data from being persisted to the database.
     # This method manually runs validations that were declared with a specific validation context (:after_import).
+    # Or runs "#{klass.name}Validator" method after_import_batch_validations for large import CSVs
     # The result is warnings are generated for the end user while the data is allowed to persist to the database.
-    def after_import_validations(records, failed_instances, options)
-      # Since row indexes start at 0 and spreadsheets on line 1,
-      # add 1 for the difference in indexes and 1 for the header row itself.
-      row_offset = CSV_FIRST_LINE + (options[:skip_lines] || 0)
+    def after_import_validations_and_row_assignment(records, failed_instances, options)
+      # this a call to custom batch validation checks for large import CSVs
+      validator_klass = "#{klass.name}Validator".safe_constantize
+      validator_klass.try(:after_import_batch_validations, failed_instances)
 
       records.each_with_index do |record, index|
-        unless record.valid?(:after_import)
-          record.errors.add(:row, "Line #{index + row_offset}")
-          failed_instances << record if record.persisted?
-        end
+        next if record.valid?(:after_import)
+
+        record.errors.add(:row, csv_row(index, options))
```

Comment:
This line is adding row numbers to all `invalid` records, whether they failed `after_import` validations or regular ActiveRecord validations or both.